aeneas/src/ssa/SsaOptimizer.v3

```diff
@@ -1963,6 +1974,154 @@ class SsaLoopOptimizer {
 			var iv = l.head;
 			if (cond != null) tryBoundingIv(iv, cond, condTrue);
 		}
+
+	}
+	def licm() {
+		if (context.shouldPrintLicm()) Terminal.putln("===== LICM =====");
+
+		if (blockOrdering == null || blockOrdering.order == null || blockOrdering.order.length <= 1) return; 
+		marker = blockOrdering.marker;
+		// Go through each block; accumulate loop depth info; and hoist invariants
+		var loopsInScope: Vector<SsaLoopInfo> = Vector<SsaLoopInfo>.new();
+		var outermostLoopBlockMark: u31;
+		for (i = 0; i < blockOrdering.order.length; i++) {
+			def curBlock: SsaBlock = blockOrdering.order[i];
+			context.block = curBlock;
+
+			def curBlockMark = (marker.getMark(curBlock));
+			for (curLink = curBlock.next; curLink != SsaLink.!(curBlock); curLink = curLink.next) {
+				marker.setMark(curLink, u31.!(curBlockMark));
+			}
+			// Case 1.) The block is the first block in the loop (header)
+			if (curBlock.info.loop != null) {
+				def loopInfo: SsaLoopInfo = curBlock.info.loop;
+				loopsInScope.put(loopInfo);
+				continue;
+			}
+			// Case 2.) The block is the exit node of the loop => need to update loopsInScope
+			while (loopsInScope.length > 0 && loopsInScope[loopsInScope.length - 1].end == i) {
+				def recentLoop: SsaLoopInfo = loopsInScope[loopsInScope.length-1];
+				loopsInScope.resize(loopsInScope.length - 1);
+			}
+			// Case 3.) The block is not in a loop 
+			if (loopsInScope.length == 0) continue;
+			// Case 4.) The block is in the body of the loop 
+			// Scan the block for invariants to hoist 
+			var nextLink: SsaLink;
+			for (link = curBlock.next; link != curBlock; link = nextLink) {
+				// Grab the next link in the current block. `link` could be hoisted elsewhere. 
+				nextLink = link.next;
+				if (link.facts.O_PURE && SsaApplyOp.?(link) &&
+					instrDominatesExitsAndBackedge(SsaInstr.!(link), curBlock, loopsInScope[loopsInScope.length - 1])) {
```

Comment:
I think we don't need to check every instruction in the block, just that the block dominates.